Add startup script library debugging functions

jeffmccune opened the pull request:
This commit adds the basic structure of the startup script library.  The
intent is to focus on logging functions and hand off of control to the
startup-script-custom script.

See examples/hello_world/ for usage information and the README for
example output.

Testing
===

BATS unit tests are being ported over, this pull request is up for early
comment.

jeffmccune committed:
Add startup script library debugging functions

This commit adds the basic structure of the startup script library.  The
intent is to focus on logging functions and hand off of control to the
startup-script-custom script.

See examples/hello_world/ for usage information and the README for
example output.

jeffmccune commented on files/startup-script-stdlib.sh:
```diff
@@ -0,0 +1,230 @@
+# Pending operational functions:
+# TODO(jmccune): get_from_bucket()
+# TODO(jmccune): setup_init_script()
+# TODO(jmccune): setup_sudoers()
```
This file contains the library of startup script functions and the TODO items are what will be added in subsequent pull requests which have already been used in the field.

jeffmccune commented on files/startup-script-stdlib.sh:
```diff
@@ -0,0 +1,230 @@
+readonly E_MISSING_MANDATORY_ARG=9
+readonly E_UNKNOWN_ARG=10
+
+stdlib::debug() {
```
I'm not sure about the package name.  Previously (in the field) there was no package name, but I added `stdlib::` based on https://google.github.io/styleguide/shell.xml?showone=Function_Names#Function_Names

> Separate libraries with ::

adrienthebo replied:
There may be a compelling reason to use a different namespace, but right now I can't think of one. My vote is that we start with `stdlib` for v1 and address name conflicts later, if ever.

jeffmccune replied:
Sounds good, will keep stdlib for now.  Thanks Adrien.

jeffmccune commented on files/startup-script-stdlib.sh:
```diff
@@ -0,0 +1,230 @@
+  DELETE_AT_EXIT="$(mktemp -d)"
+  readonly DELETE_AT_EXIT
+  # Set TMPDIR for mktemp() to clean up nicely.
+  declare -xr TMPDIR="$DELETE_AT_EXIT"
```
Exporting TMPDIR likely needs to change.  If a daemon spawns in the background and runs after this script exits, the daemon's TMPDIR will be deleted from under it which will likely cause problems.

jeffmccune replied:
This has been fixed in 9458a1d..ad57d83

adrienthebo commented on README.md:
````diff
@@ -1,2 +1,75 @@
+Inject the stdlib into the compute instance by reading the output content like
+so:
+
+```terraform
````
Is this meant to be used within a `google_compute_instance` resource? If so, can we add the context to indicate this? Something like so:

```hcl
resource "google_compute_instance" "example" {
  # other attributes removed
  metadata {
    # ...
  }
}
```

If you think this is more clear, feel free to add it; resolve this comment otherwise.

jeffmccune replied:
Much more clear, thanks for the review.  Added the example.

adrienthebo commented on files/startup-script-stdlib.sh:
```diff
@@ -0,0 +1,230 @@
+#! /bin/bash
+#
+# Copyright 2018 Google LLC. This software is provided as-is,
```
Since we're open sourcing this, I think we should use the standard [Google Apache 2 boilerplate](https://github.com/terraform-google-modules/terraform-google-project-factory/blob/master/test/boilerplate/boilerplate.sh.txt).

jeffmccune replied:
Good call.  I layered the behavior change commit on top a module cookiecutter commit and the boilerplate is passing the check now:

```
make
Running shellcheck
Running flake8
./helpers/combine_docfiles.py:32:6: W605 invalid escape sequence '\['
./helpers/combine_docfiles.py:32:8: W605 invalid escape sequence '\^'
./helpers/combine_docfiles.py:32:10: W605 invalid escape sequence '\]'
./helpers/combine_docfiles.py:32:12: W605 invalid escape sequence '\:'
./helpers/combine_docfiles.py:32:14: W605 invalid escape sequence '\ '
./helpers/combine_docfiles.py:32:16: W605 invalid escape sequence '\('
./helpers/combine_docfiles.py:32:36: W605 invalid escape sequence '\)'
./helpers/combine_docfiles.py:32:47: W605 invalid escape sequence '\['
./helpers/combine_docfiles.py:32:49: W605 invalid escape sequence '\^'
./helpers/combine_docfiles.py:32:51: W605 invalid escape sequence '\]'
./helpers/combine_docfiles.py:32:53: W605 invalid escape sequence '\:'
./helpers/combine_docfiles.py:32:55: W605 invalid escape sequence '\ '
./helpers/combine_docfiles.py:32:57: W605 invalid escape sequence '\('
./helpers/combine_docfiles.py:32:75: W605 invalid escape sequence '\)'
./helpers/combine_docfiles.py:33:19: W605 invalid escape sequence '\d'
./helpers/combine_docfiles.py:33:21: W605 invalid escape sequence '\d'
Running go fmt and go vet
Running terraform validate
Running hadolint on Dockerfiles
Checking for required files
Testing the validity of the header check
..
----------------------------------------------------------------------
Ran 2 tests in 0.051s

OK
Checking file headers
The following lines have trailing whitespace
Generating markdown docs with terraform-docs
```

jeffmccune committed:
Initialize module with cookiecutter

Without this patch the module is blank.  This patch creates the module
structure using:

cookiecutter https://partner-code.googlesource.com/cloud-foundation-toolkit.git \
  --checkout cookiecutter-template

adrienthebo commented on examples/hello_world/main.tf:
```diff
@@ -0,0 +1,52 @@
+# Your use of it is subject to your agreements with Google.
+
+provider "google" {
+  version = "~> 1.8"
```
Small nitpick: Let's bump this to 1.19 or 1.20 or so.

jeffmccune replied:
Bumped to `~> 1.20` and switched to examples/simple_example/main.tf following the cookiecutter.

adrienthebo commented on examples/hello_world/README.md:
```diff
@@ -0,0 +1,68 @@
+[jmccune@startup-script-stdlib-example1 ~]$ sudo DEBUG=1 google_metadata_script_runner --script-type startup --debug
+startup-script: INFO Starting startup scripts.
+startup-script: INFO Found startup-script in metadata.
+startup-script: INFO startup-script: 'Tue Dec 18 01:39:35 +0000 2018' Debug [5064]: init_startup_stdlib(): startup-script-stdlib.sh initialized and ready
```
Nitpick: This nesting of `startup-script: INFO startup-script: ` is a bit funny, is that intended behavior?

jeffmccune replied:
It's sort of intended.  The nesting is a result of running via the `google_metadata_script_runner` command.  It's there via $PROG to print the name when running directly, without the script runner.

Since the library will almost always run inside of the script runner, I think it makes a lot of sense to cut this out of the output.

jeffmccune replied:
Actually, I take that back.  `startup-script: INFO startup-script: ` comes from the script runner when in `--debug` mode.

adrienthebo replied:
This is a cosmetic detail, so I don't think we need to pick this over too closely - let's leave this as-is.

adrienthebo commented on files/startup-script-stdlib.sh:
```diff
@@ -0,0 +1,230 @@
+  [[ -z "${DEBUG:-}" ]] && return 0
+  local ds msg="$*"
+  logger -p "${SYSLOG_DEBUG_PRIORITY}" -t "${PROG}[$$]" -- "${msg}"
+  [[ -n "${QUIET:-}" ]] && return 0
```
We have the `DEBUG` and `QUIET` environment variables that have semantics, but I don't see documentation for user-tunable variables for this script. Can we add documentation for the variables that users may care about?

jeffmccune replied:
Good idea, added to the README.md

adrienthebo commented on files/startup-script-stdlib.sh:
```diff
@@ -0,0 +1,230 @@
+}
+
+stdlib::main() {
+  local OPTIND opt custom_startup_script
```
shellcheck is reporting that `custom_startup_script` is unused.

jeffmccune replied:
Fixed

adrienthebo reviewed (commented):
I have a smattering of small questions or suggestions but as a whole I'm happy with this. Once we have tests included in this PR I'll verify those and check the examples and give a :+1: after.

adrienthebo commented on outputs.tf:
```diff
@@ -0,0 +1,8 @@
+
+output "content" {
+  description = "startup-script-stdlib.sh content as a string value."
+  value       = "${file("${path.module}/files/startup-script-stdlib.sh")}"
```
Nitpick: Our internal standards guide recommends that scripts called by Terraform should be kept in `scripts/`. You can follow this if you like, otherwise feel free to resolve this comment.

jeffmccune replied:
I thought about this too, and it was originally in scripts/.  At the Terraform layer, the file is treated as a big string blob and passed as a resource attribute, it's never executed by Terraform directly.  Given this use case, we moved it into the files/ directory.

adrienthebo commented on files/startup-script-stdlib.sh:
```diff
@@ -0,0 +1,230 @@
+
+stdlib::debug() {
+  [[ -z "${DEBUG:-}" ]] && return 0
+  local ds msg="$*"
```
I've had shellcheck squawk at me when I declared and assigned to a local variable in the same expression. The fact that we're declaring `ds` as a local, and then declaring `msg` as a local and then assigning it is a little surprising. In terms of correctness this code is fine as-is, but I think it might be more readable to break out `ds` into a separate declaration. Since this is a style comment, feel free to change if you agree, and resolve this comment otherwise.

jeffmccune replied:
I went back and forth on this one too.  Where I settled with this code is, "make shellcheck happy and move on."

FWIW, shellcheck only seems to complain when the assignment happens in a subshell.  e.g. it'll complain about `local ds="$(date +%s)"` but it won't complain about `local msg="$*"`.

I'll go ahead and break these out since the shellguide mentions it as well.

jeffmccune replied:
Fixed

adrienthebo commented:
The addition of the `project_id` and `region` variables broke the example; I got things working with the following:

```diff
diff --git i/examples/simple_example/main.tf w/examples/simple_example/main.tf
index 10cd920..4afb5be 100644
--- i/examples/simple_example/main.tf
+++ w/examples/simple_example/main.tf
@@ -23,6 +23,9 @@ provider "google" {
 
 module "startup-scripts" {
   source = "../../"
+
+  project_id = "${var.project_id}"
+  region     = "${var.region}"
 }
 
 data "google_compute_image" "os" {
```

Do we need those variables defined on the module?

Once those changes applied the module worked as expected.

adrienthebo commented:
Currently there are two outstanding issues:

- https://github.com/terraform-google-modules/terraform-google-startup-scripts/pull/1#discussion_r242384604
- https://github.com/terraform-google-modules/terraform-google-startup-scripts/pull/1#issuecomment-448393478

:+1: following the resolution of those issues. With respect to BATS tests we can add them in this pull request, or we can follow up with another pull request (which I'd prefer over having a single monolithic PR.)

morgante commented:
+1 to adding testing in a follow-up PR

jeffmccune commented:
The example not working without project_id and region is unexpected. It should work without you needing to add them, and the module itself is intended to take no input variables so I’ll check that out first thing tomorrow morning and get it sorted.

Will also fix TMPDIR which should be straightforward.  Plan to add stdlib::mktemp with the same behavior as now, only difference is TMPDIR won’t be exported. 

Noted regarding tests, will add and follow up.

jeffmccune commented:
@adrienthebo Regarding your project_id and region comment in https://github.com/terraform-google-modules/terraform-google-startup-scripts/pull/1#issuecomment-448393478 I verified the steps as documented and fixed up in 0631423..9458a1d . Please let me know if the simple use case of `module "startup-scripts" { source = "../../" }` does not work for you.

jeffmccune commented on examples/simple_example/main.tf:
```diff
@@ -0,0 +1,64 @@
+
+module "startup-scripts" {
+  source = "../../"
+}
```
@adrienthebo Regarding your project_id and region comment in https://github.com/terraform-google-modules/terraform-google-startup-scripts/pull/1#issuecomment-448393478 I verified the steps as documented and fixed up in 0631423..9458a1d . Please let me know if the simple use case of `module "startup-scripts" { source = "../../" }` does not work for you, or resolve if it's OK now.

jeffmccune committed:
Prevent child process TMPDIR from being deleted

Without this patch the TMPDIR exported to child processes started by
the startup script library will be deleted.  This is a problem for
daemon processes requiring write access to TMPDIR.

This patch addresses the problem by not modifying TMPDIR.  Instead,
TMPDIR is temporarily overridden in the stdlib::mktemp function.  This
function should be used by library functions creating temporary files
because they'll automatically be cleaned up at exit by the trap handler.

jeffmccune commented:
> Currently there are two outstanding issues:
> 
> * [#1 (comment)](https://github.com/terraform-google-modules/terraform-google-startup-scripts/pull/1#discussion_r242384604)
> * [#1 (comment)](https://github.com/terraform-google-modules/terraform-google-startup-scripts/pull/1#issuecomment-448393478)
> 
> 👍 following the resolution of those issues. With respect to BATS tests we can add them in this pull request, or we can follow up with another pull request (which I'd prefer over having a single monolithic PR.)

Comments have been addressed in subsequent patches.  I went ahead and rebased in ad57d83...ee304af to avoid an additional round trip to clean up the commits.

adrienthebo commented:
The extra variables was an error on my part; I used the cookiecutter template to get access to the standard linters and got an extra `variables.tf` along the way.

adrienthebo reviewed (approved):
:+1: everything checks out, example works well, let's get this merged.

jeffmccune commented:
@adrienthebo FYI, spec test structure is up for review at https://github.com/terraform-google-modules/terraform-google-startup-scripts/pull/1
